File: 1380-number-of-closed-islands/1380-number-of-closed-islands.cpp
class Solution {
public:
    int dfs(int i , int j,int row,int col,vector<vector<int>>& grid,vector<vector<bool>>& vis)
    {
       if(i<0 || j<0 || i==row || j==col)
            return 0;
       if(grid[i][j]==true || vis[i][j])
            return 1;
       vis[i][j] = true;
       return min(dfs(i+1,j,row,col,grid,vis),min(dfs(i-1,j,row,col,grid,vis),min(dfs(i,j+1,row,col,grid,vis),dfs(i,j-1,row,col,grid,vis))));          
    }
    int closedIsland(vector<vector<int>>& grid) {
        int row = grid.size();
        int col = grid[0].size();
        int ans = 0;
        vector<vector<bool>> vis(row,vector<bool>(col));
        for(int i=0;i<row;i++)
        {
            for(int j =0;j<col;j++)
            {
                if(grid[i][j]==0 && vis[i][j]==false)
                { 
                    ans += dfs(i,j,row,col,grid,vis);
                }
            }
        }
        return ans;
    }
};

File: 1380-number-of-closed-islands/1380-number-of-closed-islands.py
class Solution:
    def closedIsland(self, grid: List[List[int]]) -> int:
        row , col = len(grid), len(grid[0])
        vis = set()
        ans = 0 
        def dfs(i,j):
            if i<0 or j<0 or i==row or j ==col:
                return 0
            if grid[i][j] == 1 or (i,j) in vis:
                return 1

            vis.add((i,j))
            return min(dfs(i+1,j),dfs(i-1,j),dfs(i,j-1),dfs(i,j+1))       

        for i in range(row):
            for j in range(col):
                if grid[i][j] == 0 and (i,j) not in vis:
                    ans += dfs(i,j)
        return ans                
        